README
<!-- Author's note; this was intended to become a home page at some point,
     but I'm not interested enough in grokking HTML right now, so it became
     the README instead. Most of it should be valid Markdown. -->

Announcing Movit
================

Movit is the Modern Video Toolkit, notwithstanding that anything that's
called “modern” usually isn't, and it's really not a toolkit.

Movit aims to be a _high-quality_, _high-performance_, _open-source_
library for video filters.


TL;DR, please give me download link and system demands
======================================================

OK, you need

* A C++11 compiler. GCC will do. (I haven't tried Windows, but it
  works fine on Linux and OS X, and Movit is not very POSIX-bound.)
* GNU Make.
* A GPU capable of running OpenGL 3.0 or newer. GLES3 (for mobile devices)
  will also work.
* The [Eigen 3], [FFTW3] and [Google Test] libraries. (The library itself
  does not depend on the latter, but you probably want to run the unit tests.)
  If you also have the Google microbenchmark library, you can get some
  benchmarks as well.
* The [epoxy] library, for dealing with OpenGL extensions on various
  platforms.

Movit has been tested with various Intel, AMD and NVIDIA GPUs, on Linux
and Windows. Again, most likely, GPU compatibility shouldn't be a big issue.
See below for performance estimates.


Still TL;DR, please give me the list of filters
===============================================

Blur, diffusion, FFT-based convolution, glow, lift/gamma/gain (color
correction), mirror, mix (add two inputs), luma mix (use a map to wipe between
two inputs), overlay (the Porter-Duff “over” operation), scale (bilinear and
Lanczos), sharpen (both by unsharp mask and by Wiener filters), saturation
(or desaturation), vignette, white balance, and a deinterlacer (YADIF).

Yes, that's a short list. But they all look great, are fast and don't give
you any nasty surprises. (I'd love to include denoise and
framerate up-/downconversion to the list, but doing them well are
all research-grade problems, and Movit is currently not there.)


TL;DR, but I am interested in a programming example instead
===========================================================

Assuming you have an OpenGL context already set up (either a classic OpenGL
context, a GL 3.x forward-compatible or core context, or a GLES3 context):

<code>
  using namespace movit;
  EffectChain chain(1280, 720);

  ImageFormat inout_format;
  inout_format.color_space = COLORSPACE_sRGB;
  inout_format.gamma_curve = GAMMA_sRGB;
  FlatInput *input = new FlatInput(inout_format, FORMAT_BGRA_POSTMULTIPLIED_ALPHA, GL_UNSIGNED_BYTE, 1280, 720));
  chain.add_input(input);

  Effect *saturation_effect = chain.add_effect(new SaturationEffect());
  saturation_effect->set_float("saturation", 0.7f);

  Effect *lift_gamma_gain_effect = chain.add_effect(new LiftGammaGainEffect());
  const float gain[] = { 0.8f, 1.0f, 1.0f };
  lift_gamma_gain_effect->set_vec3("gain", &gain);

  chain.add_output(inout_format, OUTPUT_ALPHA_FORMAT_POSTMULTIPLIED);
  chain.finalize();

  for ( ;; ) {
    // Do whatever you need here to decode the next frame into <pixels>.
    input->set_pixel_data(pixels);
    chain.render_to_screen();
  }
</code>


OK, I can read a bit. What do you mean by “modern”?
===================================================

Backwards compatibility is fine and all, but sometimes we can do better
by observing that the world has moved on. In particular:

* It's 2018, so people want to edit HD video.
* It's 2018, so everybody has a GPU.
* It's 2018, so everybody has a working C++ compiler.
  (Even Microsoft fixed theirs around 2003!)

While from a programming standpoint I'd love to say that it's 2018
and interlacing does no longer exist, but that's not true (and interlacing,
hated as it might be, is actually a useful and underrated technique for
bandwidth reduction in broadcast video). Movit may eventually provide
limited support for working with interlaced video; it has a deinterlacer,
but cannot currently process video in interlaced form.


What do you mean by “high-performance”?
=======================================

Today, you can hardly get a _cellphone_ without a multi-core, SIMD-capable
CPU, and a GPU. Yet, almost all open-source pixel processing I've seen
is written using straight-up single-threaded, scalar C! Clearly there is
room for improvement here, and that improvement is sorely needed.
We want to edit 1080p video, not watch slideshows.

Movit has chosen to run all pixel processing on the GPU, mostly using GLSL
fragment shaders. While “run on the GPU” does not equal “infinite speed”,
GPU programming is probably the _simplest_ way of writing highly parallel code,
and it also frees the CPU to do other things like video decoding.

Although compute shaders are supported, and can be used for speedups if
available (currently, only the deinterlacer runs as a compute shader), it is
surprisingly hard to get good performance for compute shaders for anything
nontrivial. This is also one of the primary reasons why Movit uses GLSL and
not any of the major GPU compute frameworks (CUDA and OpenCL), although it
is also important that it is widely supported (unlike CUDA) and driver quality
generally is fairly good (unlike OpenCL).

Exactly what speeds you can expect is of course highly dependent on
your GPU and the exact filter chain you are running. As a rule of thumb,
you can run a reasonable filter chain (a lift/gamma/gain operation,
a bit of diffusion, maybe a vignette) at 720p in around 30 fps on a four-year-old
Intel laptop. If you have a somewhat newer Intel card, you can do 1080p
video without much problems. And on a low-range nVidia card of today
(GTX 550 Ti), you can probably process 4K movies directly.


What do you mean by “high-quality”?
===================================

Movit aims to be high-quality in two important aspects, namely _code quality_
and _output quality_. (Unfortunately, documentation quality is not on the
list yet. Sorry.)


High-quality output?
====================

Movit works internally in linear floating-point all the way, strongly
reducing interim round-off and clipping errors. Furthermore, Movit is
(weakly) colorspace-aware. Why do colorspaces matter? Well, here's a video frame from a typical
camera, which records in Rec. 709 (the typical HDTV color space), and here's the 
same frame misinterpreted as Rec. 601 (the typical SDTV color space):

[insert picture here]

The difference might be subtle, but would you like that color cast?
Maybe you could correct for it manually, but what if it happened on output
instead of on input? And I can promise you that once we move to more
wide-gamut color spaces, like the one in Rec. 2020 (used for UHDTV), the
difference will be anything but subtle. As of [why working in linear
light matters](http://www.4p8.com/eric.brasseur/gamma.html),
others have explained it better than I can; note also that this makes Movit
future-proof when the world moves towards 10- and 12-bit color precision
(although the latter requires Movit to change from 16-bit to 32-bit floating
point, it is a simple switch). The extra power from the GPU makes all of this
simple, so do we not need to make too many concessions for the sake of speed.

Movit does not currently do ICC profiles or advanced gamut mapping;
if you have out-of-gamut colors, they will clip. Sorry.


OK, and high-quality code?
==========================

Image processing code can be surprisingly subtle; it's easy to write
code that looks right, but that makes subtle artifacts that explode
when processed further in a later step. (Or code that simply never
worked, just that nobody cared to look at the output when a given
parameter was set. I've seen that, too.)

Movit tries to counteract this by three different strategies:

* First, _look at the output_. Does it look good? Really?
  Even if you zoom in on the results? Don't settle for “meh, I'm 
  sure that's the best it can get”.
* Second, _keep things simple_. Movit does not aim for including
  every possible video effect under the sun (there are [others out there]
  that want that); the [YAGNI] principle is applied quite strongly throughout
  the code. It's much better to write less code but actually
  understand what it does; whenever I can replace some magic matrix
  or obscure formula from the web with a clean calculation and a descriptive
  comment on top, it makes me a bit happier. (Most of the time,
  it turns out that I had used the matrix or formula in a wrong
  way anyway. My degree is in multimedia signal processing, but it
  does not mean I have a deep understanding of everything people do
  in graphics.)
* Third, _have unit tests_. Tests are boring, but they are unforgiving
  (much more unforgiving than your eye), and they keep stuff from breaking
  afterwards. Almost every single test I wrote has uncovered bugs in Movit,
  so they have already paid for themselves.

There is, of course, always room for improvement. I'm sure you can find
things that are stupid, little-thought-out, or buggy. If so, please let me
know.


What do you mean by “open-source”?
==================================

Movit is licensed under the [GNU GPL](http://www.gnu.org/licenses/gpl.html),
either version 2 or (at your option) any later version. You can find the full
text of the license in the COPYING file, included with Movit.


NEWS
Movit 1.6.3, July 7th, 2019

  - Various bugfixes.


Movit 1.6.2, March 18th, 2018

  - Various bugfixes.


Movit 1.6.1, January 31st, 2018

  - Loosen up some restrictions on strong one-to-one-effects. In particular,
    this fixes a bug with MixEffect and compute shaders that hit some
    Nageru setups.


Movit 1.6.0, January 24th, 2018

  - Support for effects that work as compute shaders. Compute shaders are
    generally slower than fragment shaders for the same algorithm,
    but allow some forms of communication between shader invocations
    and have more flexible output, which can enable more efficient algorithms. 
    See effect.h for more details. Note that the fastest rendering API on
    EffectChain is now to a texture if possible, not to an FBO. This will
    only matter if the last effect is a compute shader.
    
  - Movit now includes a compute shader implementation of DeinterlaceEffect,
    which is automatically used instead of the fragment shader implementation
    if your GPU and OpenGL driver supports it (in practice, this means on
    all platforms except on macOS). The compute shader version is typically
    20–80% faster than the fragment shader version, depending on your GPU
    and other factors.

    A compute shader implementation of ResampleEffect was written but
    ultimately failed to be faster, and so is not included.

  - Support for microbenchmarks of effects through the Google microbenchmarking
    framework (optional). Currently, DeinterlaceEffect and ResampleEffect has
    benchmarks; enable them by running the unit test with --benchmark (also try
    --benchmark --help).

  - Effects can now explicitly request _not_ to have mipmaps, which means they
    can do so without needing to request bounce and fiddling with the sampler
    state. Note that this is an API change for effects.

  - Movit now requires C++11, both to build and to #include the header files.
    Support for SDL1 has been dropped; unit tests and the demo program now need
    SDL2.

  - Various smaller bugfixes and optimizations.
    

Movit 1.5.3, August 10th, 2017

  - Various bugfixes.


Movit 1.5.2, July 5th, 2017

  - The texture generation for ResampleEffect (when the size is changed)
    now uses less CPU.

  - Work around an NVIDIA driver threading bug by aggressively caching VAOs.


Movit 1.5.1, May 29th, 2017

  - Allow YCbCrInput to change input format after finalize.

  - Some minor YCbCrInput bugfixes.


Movit 1.5.0, March 21st, 2017

  - Support interleaved Y'CbCr input (4:4:4 in a single texture).

  - Support 10-bit and 12-bit Y'CbCr, both for input and output. For planar,
    these are supported packed in 16-bit ints; for interleaved, 10:10:10:2 is
    supported. (Efficient conversion to and from v210, ie. 10-bit 4:2:2,
    is possible using compute shaders, but Movit does not include support
    for them at the current point.) Note that this now means the num_levels
    flag in YCbCrFormat actually matters, although 0 will be interpreted
    as 256 (8-bit) for the benefit of older applications.

  - Limited support for having multiple Y'CbCr outputs from a chain.

  - Allow changing the Y'CbCr output coefficients runtime, ie., after finalize.

  - Fix an issue where the last pass would have been rendered with the sRGB
    flag set, which confused Qt applications running in certain NVIDIA
    configurations.


Movit 1.4.0, November 5th, 2016

  - Allow setting the intermediate format for chains, instead of hardcoding
    it at 16-bit RGBA; advanced users can use this to e.g. ask for 8-bit
    sRGB intermediates, reducing the amount of memory bandwidth needed
    at the cost of reduced precision. Whether this is a good tradeoff or not
    depends on the exact chain and your requirements.

  - Fix an issue where a (cached) shader program could be used from multiple
    threads at a time, causing the uniforms to contain unpredictable values.

  - Make the error printed on check_error() slightly friendlier: Include the
    enum if possible, and print it to stderr instead of stdout.


Movit 1.3.2, February 23rd, 2016

  - Fix an issue with initialization in certain locales. Patch from
    Jean-Baptiste Mardelle.


Movit 1.3.1, February 15th, 2016

  - Fix an issue where certain effect chains (particularly involving
    out-of-tree effects that return only a constant color) could cause
    texture coordinates not to be set properly. (The new code probably
    also has slightly lower OpenGL driver overhead.) Reported by
    Christophe Thommeret.


Movit 1.3.0, January 31st, 2016

  - Movit now requires GLSL 1.30 (so a driver from 2008 or later);
    before, it claimed to support 1.10, but actually used 1.30 features.
    Note that some OpenGL drivers, in particular on OS X, only supports
    GLSL 1.30 (actually, 1.50) if you have an OpenGL core context.

  - Add a deinterlacer, based on YADIF.

  - Allow parallel output to RGBA and Y'CbCr textures at the same time.

  - Make FlatInput and YCbCrInput support taking in external OpenGL
    textures. Also allow them to change width/height after instantiation.

  - Various compatibility and performance fixes. In particular,
    fp16 conversion on older (non-Haswell) CPUs is much faster, due to
    new conversion code by Fabian Giesen.


Movit 1.2.0, September 24th, 2015

  - Movit now ships a version.h with a #define MOVIT_VERSION that
    increases on every API change, including in git. (The number
    itself carries no semantic meaning beyond this.) Movit 1.2.0
    is not API or ABI compatible with 1.1.x, so there has been a
    soname bump to libmovit.so.3.

  - More flexible Y'CbCr input; in particular, a special input
    for 4:2:2 interleaved data (UYVY), and support for semi-planar
    input (Cb and Cr in the same texture, like in NV12). Note that
    you should now also set the new “num_levels” member of
    YCbYCbCrFormat to 256 (signifying 8-bit input), although

demo.cpp
#define NO_SDL_GLEXT 1

#define WIDTH 1280
#define HEIGHT 720

#include <epoxy/gl.h>

#include <SDL2/SDL.h>
#include <SDL2/SDL_error.h>
#include <SDL2/SDL_events.h>
#include <SDL2/SDL_image.h>
#include <SDL2/SDL_keyboard.h>
#include <SDL2/SDL_mouse.h>
#include <SDL2/SDL_video.h>

#include <assert.h>
#include <features.h>
#include <math.h>
#include <png.h>
#include <pngconf.h>
#include <setjmp.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <time.h>

#include "diffusion_effect.h"
#include "effect.h"
#include "effect_chain.h"
#include "flat_input.h"
#include "image_format.h"
#include "init.h"
#include "lift_gamma_gain_effect.h"
#include "saturation_effect.h"
#include "util.h"
#include "widgets.h"

using namespace movit;

unsigned char result[WIDTH * HEIGHT * 4];

float lift_theta = 0.0f, lift_rad = 0.0f, lift_v = 0.0f;
float gamma_theta = 0.0f, gamma_rad = 0.0f, gamma_v = 0.5f;
float gain_theta = 0.0f, gain_rad = 0.0f, gain_v = 0.25f;
float saturation = 1.0f;

//float radius = 0.3f;
// float inner_radius = 0.3f;
float blur_radius = 20.0f;
float blurred_mix_amount = 0.5f;
	
void update_hsv(Effect *lift_gamma_gain_effect, Effect *saturation_effect)
{
	RGBTriplet lift(0.0f, 0.0f, 0.0f);
	RGBTriplet gamma(1.0f, 1.0f, 1.0f);
	RGBTriplet gain(1.0f, 1.0f, 1.0f);

	hsv2rgb_normalized(lift_theta, lift_rad, lift_v, &lift.r, &lift.g, &lift.b);
	hsv2rgb_normalized(gamma_theta, gamma_rad, gamma_v * 2.0f, &gamma.r, &gamma.g, &gamma.b);
	hsv2rgb_normalized(gain_theta, gain_rad, gain_v * 4.0f, &gain.r, &gain.g, &gain.b);

	bool ok = lift_gamma_gain_effect->set_vec3("lift", (float *)&lift);
	ok = ok && lift_gamma_gain_effect->set_vec3("gamma", (float *)&gamma);
	ok = ok && lift_gamma_gain_effect->set_vec3("gain", (float *)&gain);
	assert(ok);

	if (saturation < 0.0) {
		saturation = 0.0;
	}
	ok = saturation_effect->set_float("saturation", saturation);
	assert(ok);
}

void mouse(int x, int y)
{
	float xf = (x / (float)WIDTH) * 16.0f / 9.0f;
	float yf = (HEIGHT - y) / (float)HEIGHT;

	if (yf < 0.2f) {
		read_colorwheel(xf, yf, &lift_rad, &lift_theta, &lift_v);
	} else if (yf >= 0.2f && yf < 0.4f) {
		read_colorwheel(xf, yf - 0.2f, &gamma_rad, &gamma_theta, &gamma_v);
	} else if (yf >= 0.4f && yf < 0.6f) {
		read_colorwheel(xf, yf - 0.4f, &gain_rad, &gain_theta, &gain_v);
	} else if (yf >= 0.6f && yf < 0.62f && xf < 0.2f) {
		saturation = (xf / 0.2f) * 4.0f;
#if 0
	} else if (yf >= 0.65f && yf < 0.67f && xf < 0.2f) {
		radius = (xf / 0.2f);
	} else if (yf >= 0.70f && yf < 0.72f && xf < 0.2f) {
		inner_radius = (xf / 0.2f);
#endif
	} else if (yf >= 0.75f && yf < 0.77f && xf < 0.2f) {
		blur_radius = (xf / 0.2f) * 100.0f;
	} else if (yf >= 0.80f && yf < 0.82f && xf < 0.2f) {
		blurred_mix_amount = (xf / 0.2f);
	}
}

unsigned char *load_image(const char *filename, unsigned *w, unsigned *h)
{
	SDL_Surface *img = IMG_Load(filename);
	if (img == nullptr) {
		fprintf(stderr, "Load of '%s' failed\n", filename);
		exit(1);
	}

	SDL_PixelFormat rgba_fmt;
	rgba_fmt.palette = nullptr;
	rgba_fmt.BitsPerPixel = 32;
	rgba_fmt.BytesPerPixel = 8;
	rgba_fmt.Rloss = rgba_fmt.Gloss = rgba_fmt.Bloss = rgba_fmt.Aloss = 0;

	// NOTE: Assumes little endian.
	rgba_fmt.Rmask = 0x00ff0000;
	rgba_fmt.Gmask = 0x0000ff00;
	rgba_fmt.Bmask = 0x000000ff;
	rgba_fmt.Amask = 0xff000000;

	rgba_fmt.Rshift = 16;
	rgba_fmt.Gshift = 8;
	rgba_fmt.Bshift = 0;
	rgba_fmt.Ashift = 24;

	SDL_Surface *converted = SDL_ConvertSurface(img, &rgba_fmt, SDL_SWSURFACE);

	*w = img->w;
	*h = img->h;

	SDL_FreeSurface(img);

	return (unsigned char *)converted->pixels;
}

void write_png(const char *filename, unsigned char *screenbuf)
{
	FILE *fp = fopen(filename, "wb");
	png_structp png_ptr = png_create_write_struct(PNG_LIBPNG_VER_STRING, nullptr, nullptr, nullptr);
	png_infop info_ptr = png_create_info_struct(png_ptr);
	
	if (setjmp(png_jmpbuf(png_ptr))) {
		fclose(fp);
		fprintf(stderr, "Write to %s failed; exiting.\n", filename);
		exit(1);
	}

	png_set_IHDR(png_ptr, info_ptr, WIDTH, HEIGHT, 8, PNG_COLOR_TYPE_RGB_ALPHA, PNG_INTERLACE_NONE, PNG_COMPRESSION_TYPE_DEFAULT, PNG_FILTER_TYPE_DEFAULT);

	png_bytep *row_pointers = new png_bytep[HEIGHT];
	for (unsigned y = 0; y < HEIGHT; ++y) {
		row_pointers[y] = screenbuf + ((HEIGHT - y - 1) * WIDTH) * 4;
	}

	png_init_io(png_ptr, fp);
	png_set_rows(png_ptr, info_ptr, row_pointers);
	png_write_png(png_ptr, info_ptr, PNG_TRANSFORM_BGR, nullptr);
	png_destroy_write_struct(&png_ptr, &info_ptr);
	fclose(fp);

	delete[] row_pointers;
}

int main(int argc, char **argv)
{
	bool quit = false;

	if (SDL_Init(SDL_INIT_EVERYTHING) == -1) {
		fprintf(stderr, "SDL_Init failed: %s\n", SDL_GetError());
		exit(1);
	}
	SDL_GL_SetAttribute(SDL_GL_ALPHA_SIZE, 8);
	SDL_GL_SetAttribute(SDL_GL_DEPTH_SIZE, 0);
	SDL_GL_SetAttribute(SDL_GL_STENCIL_SIZE, 0);
	SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, 1);

	SDL_GL_SetAttribute(SDL_GL_CONTEXT_PROFILE_MASK, SDL_GL_CONTEXT_PROFILE_CORE);
	SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, 3);
	SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, 1);
	// SDL_GL_SetAttribute(SDL_GL_CONTEXT_FLAGS, SDL_GL_CONTEXT_DEBUG_FLAG);
	SDL_Window *window = SDL_CreateWindow("OpenGL window",
		SDL_WINDOWPOS_UNDEFINED,
		SDL_WINDOWPOS_UNDEFINED,
		WIDTH, HEIGHT,
		SDL_WINDOW_OPENGL);
	SDL_GLContext context = SDL_GL_CreateContext(window);
	assert(context != nullptr);

	CHECK(init_movit(".", MOVIT_DEBUG_ON));
	printf("GPU texture subpixel precision: about %.1f bits\n",
		log2(1.0f / movit_texel_subpixel_precision));
	printf("Wrongly rounded x+0.48 or x+0.52 values: %d/510\n",
		movit_num_wrongly_rounded);
	if (movit_num_wrongly_rounded > 0) {
		printf("Rounding off in the shader to compensate.\n");
	}
	
	unsigned img_w, img_h;
	unsigned char *src_img = load_image(argc > 1 ? argv[1] : "blg_wheels_woman_1.jpg", &img_w, &img_h);

	EffectChain chain(WIDTH, HEIGHT);
	glViewport(0, 0, WIDTH, HEIGHT);

	ImageFormat inout_format;
	inout_format.color_space = COLORSPACE_sRGB;
	inout_format.gamma_curve = GAMMA_sRGB;

	FlatInput *input = new FlatInput(inout_format, FORMAT_BGRA_POSTMULTIPLIED_ALPHA, GL_UNSIGNED_BYTE, img_w, img_h);
	chain.add_input(input);
	Effect *lift_gamma_gain_effect = chain.add_effect(new LiftGammaGainEffect());
	Effect *saturation_effect = chain.add_effect(new SaturationEffect());
	Effect *diffusion_effect = chain.add_effect(new DiffusionEffect());
	//Effect *vignette_effect = chain.add_effect(new VignetteEffect());
	//Effect *sandbox_effect = chain.add_effect(new SandboxEffect());
	//sandbox_effect->set_float("parm", 42.0f);
	//chain.add_effect(new MirrorEffect());
	chain.add_output(inout_format, OUTPUT_ALPHA_FORMAT_POSTMULTIPLIED);
	chain.set_dither_bits(8);
	chain.finalize();

	// generate a PBO to hold the data we read back with glReadPixels()
	// (Intel/DRI goes into a slow path if we don't read to PBO)
	GLuint pbo;
	glGenBuffers(1, &pbo);
	glBindBuffer(GL_PIXEL_PACK_BUFFER_ARB, pbo);
	glBufferData(GL_PIXEL_PACK_BUFFER_ARB, WIDTH * HEIGHT * 4, nullptr, GL_STREAM_READ);

	init_hsv_resources();
	check_error();

	int frame = 0;
	bool screenshot = false;
#if _POSIX_C_SOURCE >= 199309L
	struct timespec start, now;
	clock_gettime(CLOCK_MONOTONIC, &start);
#else
	struct timeval start, now;
	gettimeofday(&start, nullptr);
#endif

	while (!quit) {
		SDL_Event event;
		while (SDL_PollEvent(&event)) {
			if (event.type == SDL_QUIT) {
				quit = true;
			} else if (event.type == SDL_KEYDOWN && event.key.keysym.sym == SDLK_ESCAPE) {
				quit = true;
			} else if (event.type == SDL_KEYDOWN && event.key.keysym.sym == SDLK_F1) {
				screenshot = true;
			} else if (event.type == SDL_MOUSEBUTTONDOWN && event.button.button == SDL_BUTTON_LEFT) {
				mouse(event.button.x, event.button.y);
			} else if (event.type == SDL_MOUSEMOTION && (event.motion.state & SDL_BUTTON(1))) {
				mouse(event.motion.x, event.motion.y);
			}
		}

		++frame;

		update_hsv(lift_gamma_gain_effect, saturation_effect);
		//vignette_effect->set_float("radius", radius);
		//vignette_effect->set_float("inner_radius", inner_radius);
		//vignette_effect->set_vec2("center", (float[]){ 0.7f, 0.5f });

		CHECK(diffusion_effect->set_float("radius", blur_radius));
		CHECK(diffusion_effect->set_float("blurred_mix_amount", blurred_mix_amount));

		input->set_pixel_data(src_img);
		chain.render_to_screen();
		
		glBindBuffer(GL_PIXEL_PACK_BUFFER_ARB, pbo);
		check_error();
		glReadPixels(0, 0, WIDTH, HEIGHT, GL_BGRA, GL_UNSIGNED_INT_8_8_8_8_REV, BUFFER_OFFSET(0));
		check_error();
		glBindBuffer(GL_PIXEL_PACK_BUFFER_ARB, 0);
		check_error();

		draw_hsv_wheel(0.0f, lift_rad, lift_theta, lift_v);
		draw_hsv_wheel(0.2f, gamma_rad, gamma_theta, gamma_v);
		draw_hsv_wheel(0.4f, gain_rad, gain_theta, gain_v);
		draw_saturation_bar(0.6f, saturation / 4.0f);
#if 0
		draw_saturation_bar(0.65f, radius);
		draw_saturation_bar(0.70f, inner_radius);
#endif
		draw_saturation_bar(0.75f, blur_radius / 100.0f);
		draw_saturation_bar(0.80f, blurred_mix_amount);

		SDL_GL_SwapWindow(window);
		check_error();

		glBindBuffer(GL_PIXEL_PACK_BUFFER_ARB, pbo);
		check_error();
		unsigned char *screenbuf = (unsigned char *)glMapBuffer(GL_PIXEL_PACK_BUFFER_ARB, GL_READ_ONLY);
		check_error();
		if (screenshot) {
			char filename[256];
			sprintf(filename, "frame%05d.png", frame);
			write_png(filename, screenbuf);
			printf("Screenshot: %s\n", filename);
			screenshot = false;
		}
		glUnmapBuffer(GL_PIXEL_PACK_BUFFER_ARB);
		check_error();
		glBindBuffer(GL_PIXEL_PACK_BUFFER_ARB, 0);
		check_error();

#if 1
#if _POSIX_C_SOURCE >= 199309L
		clock_gettime(CLOCK_MONOTONIC, &now);
		double elapsed = now.tv_sec - start.tv_sec +
			1e-9 * (now.tv_nsec - start.tv_nsec);
#else
		gettimeofday(&now, nullptr);
		double elapsed = now.tv_sec - start.tv_sec +
			1e-6 * (now.tv_usec - start.tv_usec);
#endif
		printf("%d frames in %.3f seconds = %.1f fps (%.1f ms/frame)\n",
			frame, elapsed, frame / elapsed,
			1e3 * elapsed / frame);

		// Reset every 100 frames, so that local variations in frame times
		// (especially for the first few frames, when the shaders are
		// compiled etc.) don't make it hard to measure for the entire
		// remaining duration of the program.
		if (frame == 100) {
			frame = 0;
			start = now;
		}
#endif
	}
	cleanup_hsv_resources();
	return 0; 
}
